[Hacktoberfest] Dropbox integration

HahaBill opened the pull request:
## Description

Dropbox integration with MindsDB allows users to perform SQL statements on their files (.csv, .tsv, .json, .parquet).

Furthermore, this feature also recursively finds users' desired files in Dropbox. That means users do not have to define paths in Dropbox and can easily just write the names of their files in MindsDB that they want to perform ML on.

To get the access token, go here: https://www.dropbox.com/en_GB/developers

```sql
CREATE DATABASE dropbox_datasource
WITH
    engine = 'dropbox',
    parameters = {
      "access_token": "<YOUR_ACCESS_TOKEN>"
    };
```

Fixes #9830 

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [x]   Test Location: Specify the URL or path for testing: `mindsdb/integrations/handlers/dropbox_handler/tests/test_dropbox_handler.py`
 - [x]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.


https://github.com/user-attachments/assets/8dd17656-12cc-444b-bcec-bbca0727c3ee

HahaBill commented:
I will re-do this to be an Application Handler like in S3 👍 I was following the old S3 implementation (DatabaseHandler) but now I see that it has been updated recently.

MinuraPunchihewa commented on mindsdb/integrations/handlers/dropbox_handler/tests/test_dropbox_handler.py:
```diff
@@ -0,0 +1,18 @@
+from mindsdb.integrations.handlers.dropbox_handler.dropbox_handler import DropboxHandler
+
+
+class S3HandlerTest(unittest.TestCase):
```
Let's rename this `DropboxHandlerTest`.

HahaBill replied:
Alright, I will do that!

MinuraPunchihewa reviewed (commented):
Hey @HahaBill,
Overall, this looks pretty great, but there's something I was hoping you could clear up for me. 

After taking a look at your code, I noticed that when a particular file is queried, all of the files are being listed and then the relevant file is being registered by name in your `_get_table()` method. So, if a file, say iris.csv is stored in a folder, say flowers, the relevant SELECT query would be,
``` 
SELECT * dropbox_datasource.`iris.csv`
```
Am I right about this?
If so, I think we need to change this. I think we should allow files to be queried by their full paths. In the above example, the query should be,
``` 
SELECT * dropbox_datasource.`flowers/iris.csv`
```
Otherwise, there is going to be a conflict if files with the same name are stored in different folders.

Further, I don't think it is necessary to register all of the files in the first place (listing all of the files with each query can affect performance, I feel), we can just allow them to be queried.

This has already been implemented for the [Amazon S3](https://github.com/mindsdb/mindsdb/tree/main/mindsdb/integrations/handlers/s3_handler) and [Google Cloud Storage](https://github.com/mindsdb/mindsdb/tree/main/mindsdb/integrations/handlers/gcs_handler) handlers. I think you can use these as examples.

Finally, when querying the `files` table to get a list of files in the Dropbox account, we would want the content of the files to be returned in a separate column called `content` . This column should be returned ONLY IF it is requested when running the `SELECT` query. This has, again, been implemented in the above handlers and you can use them as references.

ZoranPandovski commented on mindsdb/integrations/handlers/dropbox_handler/__about__.py:
```diff
@@ -0,0 +1,9 @@
+__author__ = "Ton Hoang Nguyen (Bill)"
+__github__ = "https://github.com/mindsdb/mindsdb"
+__pypi__ = "https://pypi.org/project/mindsdb/"
+__license__ = "GPL-3.0"
```
This should be MIT

HahaBill replied:
@ZoranPandovski Alright, I will change that :)

Just a note in case you are unaware: in the documentation here: https://docs.mindsdb.com/contribute/app-handlers, the GitHub Handler's license is GPL 3.0. However, when I checked the GitHub Handler in the mindsdb repository, it was MIT.

### The `__about__.py `example in the docs on mindsdb website:

<img width="605" alt="Screenshot 2024-10-28 at 20 47 38" src="https://github.com/user-attachments/assets/c21b9fae-c5b1-4316-92b8-a84893eec213">


### The `__about__.py` in the GitHub handler in the mindsdb repository (main branch):

<img width="834" alt="Screenshot 2024-10-28 at 20 50 09" src="https://github.com/user-attachments/assets/d676bbbf-551f-4fb1-9b96-7db296e6bda2">

ZoranPandovski commented on mindsdb/integrations/handlers/dropbox_handler/dropbox_handler.py:
```diff
@@ -0,0 +1,207 @@
+            return
+        if "access_token" not in self.connection_data:
+            raise ValueError("Access token must be provided.")
+        self.dbx = dropbox.Dropbox(self.connection_data["access_token"])
```
Maybe wrap this in try/except to catch some specific Auth errors?

HahaBill replied:
Alright, I will do that!

ZoranPandovski commented on mindsdb/integrations/handlers/dropbox_handler/dropbox_handler.py:
```diff
@@ -0,0 +1,207 @@
+        return files
+
+    def _read_file(self, path) -> pd.DataFrame:
+        _, res = self.dbx.files_download(path)
```
Maybe add try/except here too

HahaBill replied:
Also for this one

ZoranPandovski commented on mindsdb/integrations/handlers/dropbox_handler/dropbox_handler.py:
```diff
@@ -0,0 +1,207 @@
+        else:
+            raise ValueError(f"Unsupported file format: {extension}")
+        buffer.seek(0)
+        self.dbx.files_upload(
```
Maybe add try/except here too

HahaBill replied:
aaand also for this one

ZoranPandovski commented on mindsdb/integrations/handlers/dropbox_handler/dropbox_handler.py:
```diff
@@ -0,0 +1,207 @@
+            raise ValueError(f"Unsupported file format: {extension}")
+        buffer.seek(0)
+        self.dbx.files_upload(
+            buffer.read(), path, mode=dropbox.files.WriteMode.overwrite
```
Should we reset the buffer before read?

HahaBill replied:
I think the buffer should already be resetted (not sure if "resetted" is the correct English :p) before reading with `buffer.seek(0)`. I might be missing something but I can look into it more.

ZoranPandovski reviewed (changes requested):
Thanks, I've added few minor comments

HahaBill commented:
> Hey @HahaBill, Overall, this looks pretty great, but there's something I was hoping you could clear up for me.
> 
> After taking a look at your code, I noticed that when a particular file is queried, all of the files are being listed and then the relevant file is being registered by name in your `_get_table()` method. So, if a file, say iris.csv is stored in a folder, say flowers, the relevant SELECT query would be,
> 
> ```
> SELECT * dropbox_datasource.`iris.csv`
> ```
> 
> Am I right about this? If so, I think we need to change this. I think we should allow files to be queried by their full paths. In the above example, the query should be,
> 
> ```
> SELECT * dropbox_datasource.`flowers/iris.csv`
> ```
> 
> Otherwise, there is going to be a conflict if files with the same name are stored in different folders.
> 
> Further, I don't think it is necessary to register all of the files in the first place (listing all of the files with each query can affect performance, I feel), we can just allow them to be queried.
> 
> This has already been implemented for the [Amazon S3](https://github.com/mindsdb/mindsdb/tree/main/mindsdb/integrations/handlers/s3_handler) and [Google Cloud Storage](https://github.com/mindsdb/mindsdb/tree/main/mindsdb/integrations/handlers/gcs_handler) handlers. I think you can use these as examples.
> 
> Finally, when querying the `files` table to get a list of files in the Dropbox account, we would want the content of the files to be returned in a separate column called `content` . This column should be returned ONLY IF it is requested when running the `SELECT` query. This has, again, been implemented in the above handlers and you can use them as references.

Hi Minura! Thank you for reviewing the code and I am happy to answer all of your questions! I will make the changes according to your points. I will divide my reply into sections so it is more readable.

### 1. Listing all files when querying

The reason why is the way it is implemented is that users do not have to define specific paths (which I know that you understand) and **especially they might move files to other folders**. Maybe the latter reason is an edge case that might not occur that often. But I agree that if users have thousands of folders, the performance will degrade.

### 2. Two or more files have the same name

This is a very good catch! This edge case appeared on my mind but I forgot to act upon it. It is a good idea what you suggested: allowing users to also input the full/partial paths in the **select** query: 
```sql
SELECT * FROM dropbox_datasource.`flowers/iris.csv`
```

I will think of something how to handle this.

### 3. Registering all the files

This relates to the previous point I made about users moving files to different folders. 

Overall, I will make the changes according to your inputs and make sure to follow the pattern from S3 or Google Storage handlers :)

MinuraPunchihewa commented:
@HahaBill Sounds good, thank you. Yes, I think following the structure of these integrations is a good way to go about this.

HahaBill commented:
Hi! I've just finished making the requested changes and some improvements. I created bullet points of what is done and what is still pending.

**DONE**: 

- [x] Added ```try-except``` blocks
- [x] Allowing users to query the actual paths in Dropbox while preserving the implementation where users do not have to define the paths.
   Basically, it first checks whether the `self.dropbox_path` is a path in Dropbox or only a basename (a filename with an extension). If neither is found, it throws an exception and creates logs.
       - **path**: immediately downloads the file
       - **basename**: search in all folders and find the file
- [x] Added content column to files table when requested (```SELECT content from dropbox_datasource.files```)
- [x] Rewrote a bit of the `query` function to follow the structure

**PENDING**:
- [ ]  Discussion about resetting the buffer before reading

**DEMO:**

https://github.com/user-attachments/assets/791b5cf2-3e1e-4b31-94c8-b043c243f940

MinuraPunchihewa reviewed (changes requested):
About your comment here:
> Allowing users to query the actual paths in Dropbox while preserving the implementation where users do not have to define the paths.

I think we should remove the second option. Firstly, it does not make sense to be able to query the same file using 2 different paths: the full path and the base name.

Further, based on how you have mentioned that base names will be handled (searching in all folders to find the file); if multiple files with the same name are available in different folders, then this could cause an issue.

HahaBill commented:
> About your comment here:
> 
> > Allowing users to query the actual paths in Dropbox while preserving the implementation where users do not have to define the paths.
> 
> I think we should remove the second option. Firstly, it does not make sense to be able to query the same file using 2 different paths: the full path and the base name.
> 
> Further, based on how you have mentioned that base names will be handled (searching in all folders to find the file); if multiple files with the same name are available in different folders, then this could cause an issue.

Alright, I got it! Maybe in the future, we can handle the problem by comparing partial paths but there we might bumped into some unknown edge cases. I think for now is best to  keep it simple. I will remove the second option!

HahaBill commented on mindsdb/integrations/handlers/dropbox_handler/dropbox_handler.py:
```diff
@@ -0,0 +1,271 @@
+        table = self._get_table(Identifier(table_name))
+        columns = table.get_columns()
+        df = pd.DataFrame(columns, columns=["column_name"])
+        df["data_type"] = "string"
```
Alright, got it!

HahaBill replied:
Done!

HahaBill commented:
> > About your comment here:
> > > Allowing users to query the actual paths in Dropbox while preserving the implementation where users do not have to define the paths.
> > 
> > 
> > I think we should remove the second option. Firstly, it does not make sense to be able to query the same file using 2 different paths: the full path and the base name.
> > Further, based on how you have mentioned that base names will be handled (searching in all folders to find the file); if multiple files with the same name are available in different folders, then this could cause an issue.
> 
> Alright, I got it! Maybe in the future, we can handle the problem by comparing partial paths but there we might bumped into some unknown edge cases. I think for now is best to keep it simple. I will remove the second option!

Done! Let me know if I need to make more changes. I also tested the code, and it works without issues.

HahaBill commented:
Wait I need to check something
